### PR TITLE
Feature/init-design

### PR DESCRIPTION
**Pull request by hcheng826**
WIP, open PR to have in-line comments to discuss

### CODE REVIEW

**Review by JaisPiyush — CHANGES_REQUESTED** [review]
Remove the oracle contract completely and LoadAgent, and LoanAgentFactory should be finalized names

**Review by JaisPiyush — APPROVED** [review]

File: src/LoanAgentFactory.sol [hunk 1]
```diff
@@ -2,7 +2,34 @@
 pragma solidity ^0.8.13;
 
 contract LoadAgentFactory {
+    enum LoanRequestStatus{ PENDING, APPROVED, REJECTED }
+    mapping(address => bool) registeredLoanAgents;
+
+    modifier onlyOracle {
+        _;
```
Comment (JaisPiyush) [inline src/LoanAgentFactory.sol:35]: Add oracle contract address
```solidity
address public immutable oracle;
```
and check msg.sender is equal to this address in the modifier

File: src/LoanAgentFactory.sol [hunk 2]
```diff
@@ -2,7 +2,34 @@
 pragma solidity ^0.8.13;
 
 contract LoadAgentFactory {
+    enum LoanRequestStatus{ PENDING, APPROVED, REJECTED }
```
Comment (JaisPiyush) [inline src/LoanAgentFactory.sol]: LoanAgentFactory which manages all the LoanAgent. This enum has no use.

File: src/LoanAgent.sol [hunk 3]
```diff
@@ -2,34 +2,61 @@
...
+    enum NodeStatus {
         InQueue,
-        // Stake pool is waiting for the node to transfer the onwerhsip address to StakePool
-        // with updated quota and epoch
-        OwnershipConfirmationPending,
-        // Stake Pool is waiting for the node to host an oracle and confirm its liveliness by responding to a challenge
-        OracleConfirmationPending,
-        // Stake Pool is active
+        WaitingForOwnerChange,
+        WaitingForOracleVerifiction,
```
Comment (JaisPiyush) [inline src/LoanAgent.sol:46]: Add another status `WaitingForPleding`. This status indicated that all the validations are done and only the node is required to call pledge command from the cli.

Comment (JaisPiyush) [reply src/LoanAgent.sol:46]: I'll add the status in the issue

File: src/LoanAgent.sol [hunk 4]
```diff
@@ -40,9 +67,55 @@ contract LoadAgent {
...
+        uint256 checkpoint,
+        uint256 minerShare,
+        uint256 usersShare,
+        uint256 oracleShare
+    ) external onlyOracleContract {}
+
+    // The amount that realOwner can withdraw anytime.
+    // The amount must be less than totalOwnerCredit and available balance.
+    // The withdrawal will update totalOwnerCredit, availableBalanceHistory
+    function withdrawMinerEarning(uint256 amount) external onlyRealOwner {}
```
[the remark this answers is not in the document]
Comment (JaisPiyush) [reply src/LoanAgent.sol:135]: But msg.sender will be an `address` let us confirm the address spec first.

**Review by JaisPiyush — CHANGES_REQUESTED** [review]

File: src/LoanAgentFactory.sol (hunk 1 above)
Comment (hcheng826) [reply src/LoanAgentFactory.sol:35]: just a placeholder to make it compile, will impl later

File: src/LoanAgent.sol (hunk 3 above)
Comment (hcheng826) [reply src/LoanAgent.sol:46]: updated

File: src/aFIL.sol [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    struct Loan {
+        LoanAgent loanAgent;
+        uint totalAmount;
+        uint expectFinishTime;
+        uint remainingAmount;
+    }
+
+    mapping (address => Loan) loans;
+
+    modifier onlyOracle() {
```
Comment (JaisPiyush) [inline src/aFIL.sol]: Oracle should not update the stat of aFIL directly. Instead LoanAgentFactory should have these privileges. replace `onlyOracle` with `onlyLoanAgentFactory`

**Review by JaisPiyush — CHANGES_REQUESTED** [review]

File: src/aFIL.sol [hunk 6]
```diff
@@ -1,25 +1,74 @@
...
+
+    constructor(address _oracle) ERC20("arc FIL", "aFIL") {
+        oracle = _oracle;
     }
 
     struct Loan {
         LoanAgent loanAgent;
-        uint loanAmount;
-        uint finishTime;
+        uint totalAmount;
```
Comment (JaisPiyush) [inline src/aFIL.sol]: aFIL contract will only manage aFIL:FIL pool and should not have any kind of relation with LoanAgents or other things.

Loan and loans should be moved to LoanAgentFactory, as it holds power to create and manage them.

Comment (JaisPiyush) [reply src/aFIL.sol]: Only maintain the total loan amount

**Review by JaisPiyush — CHANGES_REQUESTED** [review]

**Review by JaisPiyush — APPROVED** [review]
Merging this pull request in main. Then I will begin working on the Oracle contract and LoanAgentFactory contract (oracle side)

**Pull request merged**